working/0508 - strict-inference/strict-inference.md

```diff
@@ -0,0 +1,362 @@
+# Strict inference static analysis option
+
+This document specifies the "Strict inference" mode enabled with a static
+analysis option. As a static analysis option, we only intend to implement this
+feature in the Dart Analyzer. Under this feature, when there is not enough
+information available to infer an expression's type, where inference "falls
+back" to `dynamic` (or the type's bound), inference is considered to have
+failed, and an analyzer "Hint" is reported at the location of the expression.
+
+## Enabling strict inference
+
+To enable strict inference, set the `strict-inference` option to `true`, under
+the Analyzer's  `language` section:
+
+```yaml
+analyzer:
+  langauge:
+    strict-inference: true
+```
+
+## Motivation
+
+It is possible to write Dart code that passes all static analysis and
+compile-time checks that is guaranteed to result in runtime errors. Commone
+examples include runtime type errors, and no-such-method errors. New and
+experienced Dart developers alike write such code, and are surprised to see
+such errors at runtime, which look like they should be caught at compile time.
+
+The strict inference mode aims to highlight such code during static analysis.
+We can look at two common examples:
+
+```dart
```

Comment:
This example doesn't work.  Downwards inference does fill in the type for [] in this example.